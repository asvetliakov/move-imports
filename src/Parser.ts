import * as ts from "typescript";

export interface LineAndCharacter {
    line: number;
    character: number;
}

export interface Reference {
    /**
     * Reference path
     */
    path: string;
    /**
     * Start position
     */
    start: LineAndCharacter;
    /**
     * End position
     */
    end: LineAndCharacter;
}

/**
 * Simple TS file parser to extract imports/exports and require-like expressions
 */
export class Parser {
    public constructor(
        private expressionReferences: string[],
    ) { }

    /**
     * Return references to other source from given source
     */
    public getReferences(source: string): Reference[] {
        const refs: Reference[] = [];

        try {
            const tsSourceFile = ts.createSourceFile("", source, ts.ScriptTarget.Latest, true, ts.ScriptKind.TSX);
            const visitor = (node: ts.Node): void => {
                if (
                    (ts.isImportDeclaration(node) || (ts.isExportDeclaration(node) && !node.exportClause)) &&
                    node.moduleSpecifier && ts.isStringLiteral(node.moduleSpecifier)) {
                    // ES6 imports/ES6 re-export
                    // !node.exportClause will be undefined for re-export
                    refs.push({
                        path: node.moduleSpecifier.text,
                        start: ts.getLineAndCharacterOfPosition(tsSourceFile, node.moduleSpecifier.getStart()),
                        end: ts.getLineAndCharacterOfPosition(tsSourceFile, node.moduleSpecifier.getEnd()),
                    });
                } else if (ts.isVariableStatement(node)) {
                    // const a = require("b"), { c } = require("c");
                    node.declarationList.declarations.forEach((decl) => {
                        if (decl.initializer && ts.isCallExpression(decl.initializer) && this.shouldProcessCallExpression(decl.initializer)) {
                            // first arg is string was asserted in this.shouldProcessCallExpression
                            const refNode = decl.initializer.arguments[0] as ts.StringLiteral;
                            refs.push({
                                path: refNode.text,
                                start: ts.getLineAndCharacterOfPosition(tsSourceFile, refNode.getStart()),
                                end: ts.getLineAndCharacterOfPosition(tsSourceFile, refNode.getEnd()),
                            });
                        }
                    });

                } else if (ts.isExpressionStatement(node)) {
                    // expression calls without variable assignment:
                    // require("./a");
                    // jest.mock("./a");
                    if (ts.isCallExpression(node.expression) && this.shouldProcessCallExpression(node.expression)) {
                        // first arg is string was asserted in this.shouldProcessCallExpression
                        const refNode = node.expression.arguments[0] as ts.StringLiteral;
                        refs.push({
                            path: refNode.text,
                            start: ts.getLineAndCharacterOfPosition(tsSourceFile, refNode.getStart()),
                            end: ts.getLineAndCharacterOfPosition(tsSourceFile, refNode.getEnd()),
                        });
                    }
                }
            };

            // Process only top-level file statements, no need to walk through full AST
            // however dynamic imports and require-like calls won't be processed if they-re not top-level
            // not a really issue i think
            tsSourceFile.statements.forEach(visitor);
            return refs;
        } catch {
            return refs;
        }
    }

    /**
     * Return true if we need to process given call expression further.
     * This will check expression identifier against this.expressionReferences array
     *
     * @private
     * @param expression
     * @returns
     */
    private shouldProcessCallExpression(expression: ts.CallExpression): boolean {
        // this will give full expression name, such as require, jest.mock, a.b.c, etc...
        const identifier = expression.expression.getText();
        // constraint that call expression should have module path as first argument
        const firstArg = expression.arguments[0];
        return (
            this.expressionReferences.includes(identifier) &&
            firstArg &&
            (ts.isStringLiteral(firstArg) || ts.isNoSubstitutionTemplateLiteral(firstArg))
        );
    }

}
