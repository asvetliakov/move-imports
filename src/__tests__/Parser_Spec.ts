import { Parser } from "../Parser";

let parser: Parser;

beforeEach(() => {
    parser = new Parser(["require", "require.actual", "jest.mock"]);
});

it("returns empty reference array if source is empty", () => {
    expect(parser.getReferences("")).toEqual([]);
});

it("returns references for es6 imports and re-exports", () => {
    const source = `
    import a from "./a";
    import { b, c } from "b";
    import * as c from "../c";
    import { default as d } from "d/test";

    export class A {}
    const B = {};

    export * from "./e";
    export { B };
    `;
    expect(parser.getReferences(source)).toMatchSnapshot();
});

it("Returns references for require-like calls if they are defined as expression references", () => {
    const source = `
    require("a");
    const b = require("b");
    const { c } = require("../c");
    const d = require("d"),
        { e } = require("./e");

    require.actual("aa");
    require.nonactual("ab");
    const f = another("k");

    jest.mock("a");
    jest.dontMock("b");
    `;
    expect(parser.getReferences(source)).toMatchSnapshot();
});

it("Skips syntax errors", () => {
    const source = `
    import a from "./a";
    impor { b } from "./b";
    import * as from "./c";
    requir(;;

    export * from "../aa";
    `;
    expect(parser.getReferences(source)).toMatchSnapshot();
});