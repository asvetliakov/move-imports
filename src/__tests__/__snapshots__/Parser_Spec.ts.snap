// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Returns references for require-like calls if they are defined as expression references 1`] = `
Array [
  Object {
    "end": Object {
      "character": 15,
      "line": 1,
    },
    "path": "a",
    "start": Object {
      "character": 12,
      "line": 1,
    },
  },
  Object {
    "end": Object {
      "character": 25,
      "line": 2,
    },
    "path": "b",
    "start": Object {
      "character": 22,
      "line": 2,
    },
  },
  Object {
    "end": Object {
      "character": 32,
      "line": 3,
    },
    "path": "../c",
    "start": Object {
      "character": 26,
      "line": 3,
    },
  },
  Object {
    "end": Object {
      "character": 25,
      "line": 4,
    },
    "path": "d",
    "start": Object {
      "character": 22,
      "line": 4,
    },
  },
  Object {
    "end": Object {
      "character": 29,
      "line": 5,
    },
    "path": "./e",
    "start": Object {
      "character": 24,
      "line": 5,
    },
  },
  Object {
    "end": Object {
      "character": 23,
      "line": 7,
    },
    "path": "aa",
    "start": Object {
      "character": 19,
      "line": 7,
    },
  },
  Object {
    "end": Object {
      "character": 17,
      "line": 11,
    },
    "path": "a",
    "start": Object {
      "character": 14,
      "line": 11,
    },
  },
]
`;

exports[`Skips syntax errors 1`] = `
Array [
  Object {
    "end": Object {
      "character": 23,
      "line": 1,
    },
    "path": "./a",
    "start": Object {
      "character": 18,
      "line": 1,
    },
  },
  Object {
    "end": Object {
      "character": 26,
      "line": 3,
    },
    "path": "./c",
    "start": Object {
      "character": 21,
      "line": 3,
    },
  },
  Object {
    "end": Object {
      "character": 25,
      "line": 6,
    },
    "path": "../aa",
    "start": Object {
      "character": 18,
      "line": 6,
    },
  },
]
`;

exports[`returns references for es6 imports and re-exports 1`] = `
Array [
  Object {
    "end": Object {
      "character": 23,
      "line": 1,
    },
    "path": "./a",
    "start": Object {
      "character": 18,
      "line": 1,
    },
  },
  Object {
    "end": Object {
      "character": 28,
      "line": 2,
    },
    "path": "b",
    "start": Object {
      "character": 25,
      "line": 2,
    },
  },
  Object {
    "end": Object {
      "character": 29,
      "line": 3,
    },
    "path": "../c",
    "start": Object {
      "character": 23,
      "line": 3,
    },
  },
  Object {
    "end": Object {
      "character": 41,
      "line": 4,
    },
    "path": "d/test",
    "start": Object {
      "character": 33,
      "line": 4,
    },
  },
  Object {
    "end": Object {
      "character": 23,
      "line": 9,
    },
    "path": "./e",
    "start": Object {
      "character": 18,
      "line": 9,
    },
  },
]
`;
